divi/src/miner.cpp

```diff
@@ -31,39 +30,39 @@ extern Settings& settings;
 extern CCriticalSection cs_main;
 extern CTxMemPool mempool;
 
-typedef std::map<unsigned int, unsigned int> LastExtensionTimestampByBlockHeight;
-LastExtensionTimestampByBlockHeight& getLastExtensionTimestampByBlockHeight()
+namespace
 {
-    static std::map<unsigned int, unsigned int> mapHashedBlocks;
-    return mapHashedBlocks;
-}
 
-std::unique_ptr<CoinMintingModule> coinMintingModule(nullptr);
+std::unique_ptr<CoinMintingModule> coinMintingModule;
+
+} // anonymous namespace
+
 void InitializeCoinMintingModule(I_StakingWallet* pwallet)
 {
-    static const CSporkManager& sporkManager = GetSporkManager();
-    static const CChainParams& chainParameters = Params();
-    static const MasternodeModule& masternodeModule = GetMasternodeModule();
-    static const I_PeerBlockNotifyService& peerNotification = GetPeerBlockNotifyService();
-    static const auto& chainstate = ChainstateManager::Get();
+    assert(coinMintingModule == nullptr);
     coinMintingModule.reset(
         new CoinMintingModule(
             settings,
             cs_main,
-            chainParameters,
-            chainstate,
-            masternodeModule,
+            Params(),
+            GetMasternodeModule(),
             FeeAndPriorityCalculator::instance().getMinimumRelayFeeRate(),
             mempool,
-            peerNotification,
+            GetPeerBlockNotifyService(),
             *pwallet,
-            getLastExtensionTimestampByBlockHeight(),
-            sporkManager));
+            GetSporkManager()));
 
 }
+
+void DestructCoinMintingModule()
+{
+    assert(coinMintingModule != nullptr);
+    coinMintingModule.reset();
```

Comment:
Add a mutex lock to both the `DestructCoinMintingModule` and before the `while(true){...}` loop inside `MinterThread`. That way you can ensure that the destruction of the minting module happens after the last reference to the `CoinMinter` instance is released. Otherwise it's presumably possible that with sufficiently many utxos, the `CoinMinter` instance is in use during deletion

Reply:
Good idea.  I think the lock belongs to `ThreadCoinMinter`, though (the function one step up in the call stack from the `while` loop), as that is the place that actually calls `GetCoinMintingModule`.  I've added those locks now.